src/monotonic_path_preview/mod.rs

```diff
@@ -0,0 +1,911 @@
+use std::io;
+use std::path;
+
+/// An error encountered during path handling.
+#[derive(Debug)]
+pub enum Error {
+    /// An error returned by the operating system.
+    ///
+    /// `err` is the underlying error returned by the operating system.
+    ///
+    /// `at` is the path that provoked the error.
+    IoError {
+        err: io::Error,
+        at: path::PathBuf,
+    },
+
+    /// Returned by [`Absolute::new()`] when given a path that involves a
+    /// symlink loop.
+    ///
+    /// [`Absolute::new()`]: struct.Absolute.html#method.new
+    SymlinkLoops(path::PathBuf),
+
+    /// Returned by [`Relative::new()`] when given a (partially or fully) absolute
+    /// path.
+    ///
+    /// [`Relative::new()`]: struct.Relative.html#method.new
+    PathIsAbsolute(path::PathBuf),
+
+    /// Returned by [`Relative::new()`] when given a path with enough `/../`
+    /// components to escape whatever prefix it's joined to.
+    ///
+    /// [`Relative::new()`]: struct.Relative.html#method.new
+    RelativePathEscapesPrefix(path::PathBuf),
+
+    #[doc(hidden)]
+    __NonExhaustive,
+}
+
+impl fmt::Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter) -> Result<(), fmt::Error> {
+        match self {
+            Error::IoError { err, at } => write!(f, "{}: {:?}", err, at),
+            Error::SymlinkLoops(p) => {
+                write!(f, "Found an infinite symlink loop: {:?}", p)
+            }
+            Error::PathIsAbsolute(p) => write!(
+                f,
+                "Tried to make a relative path from absolute path {:?}",
+                p
+            ),
+            Error::RelativePathEscapesPrefix(p) => write!(
+                f,
+                "Tried to make a relative path with leading '..': {:?}",
+                p
+            ),
+            _ => write!(f, "Unknown error"),
+        }
+    }
+}
+
+impl error::Error for Error {
+    fn description(&self) -> &str {
+        match self {
+            Error::IoError { err: e, at: _ } => e.description(),
+            Error::SymlinkLoops(_) => "Found an infinite symlink loop",
+            Error::PathIsAbsolute(_) => {
+                "Tried to make a relative path from an absolute path"
+            }
+            Error::RelativePathEscapesPrefix(_) => {
+                "Tried to make a relative path with leading '..'"
+            }
+            _ => "Unknown error",
+        }
+    }
+}
+
+/// Splits a path into a head and a tail.
+///
+/// The "head" consists of the Prefix and RootDir components, if any,
+/// and is fully canonicalized.
+///
+/// The "tail" consists of all the other components that follow the head.
+///
+/// # Errors
+///
+/// Returns [`Error::IoError`] if the head cannot be canonicalized.
+/// For example, if the process' current working directory has been deleted,
+/// or the given path includes a syntactically invalid prefix.
+///
+/// [`Error::IoError`]: struct.Error.html#variant.IoError
+fn split_head_and_tail<P: AsRef<path::Path>>(
```

Comment:
this seems like it would be generally useful!

Reply:
It does seem useful, although I can't think of any concrete use-cases off the top of my head. In this PR, however, I only wanted to demonstrate the `Absolute` and `Relative` structs, though, so I left this as an implementation detail rather than proposing it as a new API.

Feel free to add it to `path_abs` independently, though, if it takes your fancy!